src/main/scala/io/iohk/ethereum/network/discovery/Secp256k1SigAlg.scala

```diff
@@ -0,0 +1,180 @@
+package io.iohk.ethereum.network.discovery
+
+import akka.util.ByteString
+import io.iohk.ethereum.crypto
+import io.iohk.ethereum.crypto.ECDSASignature
+import io.iohk.scalanet.discovery.crypto.{SigAlg, PublicKey, PrivateKey, Signature}
+import io.iohk.ethereum.nodebuilder.SecureRandomBuilder
+import scodec.bits.BitVector
+import scodec.{Attempt, Err}
+import scodec.bits.BitVector
+
+object Secp256k1SigAlg extends SigAlg with SecureRandomBuilder {
+  override val name = "secp256k1"
+
+  override val PrivateKeyBytesSize = 32
+
+  // A Secp256k1 public key is 32 bytes compressed or 64 bytes uncompressed,
+  // with a 1 byte prefix showing which version it is.
+  // See https://davidederosa.com/basic-blockchain-programming/elliptic-curve-keys
+  //
+  // However in the discovery v4 protocol the prefix is omitted.
+  override val PublicKeyBytesSize = 64
+
+  // A normal Secp256k1 signature consists of 2 bigints `r` and `s` followed by a recovery ID `v`,
+  // but it can be just 64 bytes if that's omitted, like in the ENR.
+  override val SignatureBytesSize = 65
+
+  val SignatureWithoutRecoveryBytesSize = 64
+  val PublicKeyCompressedBytesSize = 33
+
+  override def newKeyPair: (PublicKey, PrivateKey) = {
+    val keyPair = crypto.generateKeyPair(secureRandom)
+    val (privateKeyBytes, publicKeyBytes) = crypto.keyPairToByteArrays(keyPair)
+
+    val publicKey = toPublicKey(publicKeyBytes)
+    val privateKey = toPrivateKey(privateKeyBytes)
+
+    publicKey -> privateKey
+  }
+
+  override def sign(privateKey: PrivateKey, data: BitVector): Signature = {
+    val message = crypto.kec256(data.toByteArray)
+    val keyPair = crypto.keyPairFromPrvKey(privateKey.toByteArray)
```

Comment:
We can add caching to this class as well, which would require less changes to the overall arrangement of now keys are passed around. But if by multiplying points you mean specifically that the public key has to be derived from the private ever time, I agree, it's unnecessary, the only reason I'm doing it is because `sign` takes an `AsymmetricCipherKeyPair`, but it only ever uses `keyPair.getPrivate`. So if parsing the bytes itself isn't too expensive, then I can just add another `sign` overload that takes the private key only, rather than the key pair, to avoid needlessly constructing both. Just didn't want to mess around with `crypto` to start with.

Reply:
Ah, that's not true, it uses the public key for the calculation of `v`.

Reply:
Turned into a class and added caching.